### DISCUSSION TITLE
read_holding registers

### DISCUSSION THREAD

**Discussion by jagonzalezga** [eModbus/eModbus#192] [Q&A] [answered]
Hi, I'm new to using the modbus protocol. My case study in particular is the following.

I have a lpgmass flow meter, 

I have been reading the documentation a bit to read a specific register of this device but in the code that I make it does not return anything

![Capture](https://user-images.githubusercontent.com/24602056/170587941-106920df-46b7-4f5e-8d75-6a6a7ecb233c.PNG)

first things first, the register I need to read is 4912 and it is supposed to return an integer value between 0 and 7 and i know that the value that I need to return is 7 because that is how I configured it with an external program to 115200,So my problem is that when I do the code with some of the functions of this library, it does not return the desired value.

I attach my circuit which is an esp32-wroom-32u and a max485 through uart
![WhatsApp Image 2022-05-26 at 4 03 15 PM](https://user-images.githubusercontent.com/24602056/170588673-d1b75f2d-8bda-4380-9327-c98f1028c930.jpeg)

and my code is the following.-
```
#include <Arduino.h>

// Include the header for the ModbusClient RTU style
#include "ModbusClientRTU.h"
#define rtsPin  22
// Create a ModbusRTU client instance
// In my case the RS485 module had auto halfduplex, so no second parameter with the DE/RE pin is required!
//ModbusClientRTU MB(Serial2);
ModbusClientRTU MB(Serial2, rtsPin);  // use rtsPin to toggle DE/RE in half-duplex

// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the message data and length of it, 
// plus a user-supplied token to identify the causing request
void handleData(ModbusMessage response, uint32_t token) 
{ uint16_t addr,words;
  int index=4912,index2=4912;
  
  uint16_t res1 = response.get(index,addr);
  uint16_t res2 = response.get(index2,words);
  Serial.printf("Respuesta 1=%u\n",res1);
  Serial.printf("Respuesta 2=%u\n",res2);
  Serial.printf("addr=%u\n",addr);
  Serial.printf("words=%u\n",words);
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d: \n", response.getServerID(), response.getFunctionCode(), token, response.size());
  for (auto& byte : response) {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", (int)me, (const char *)me);
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Set up Serial2 connected to Modbus RTU
// (Fill in your data here!)
  Serial2.begin(115200, SERIAL_8E1, GPIO_NUM_17, GPIO_NUM_18);

// Set up ModbusRTU client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task
  MB.begin();




// If something is missing or wrong with the call parameters, we will immediately get an error code 
// and the request will not be issued
  uint32_t Token = 1111;


  Error err;


// Create request for
// (Fill in your data here!)
// - server ID = 1  //masico 247
// - function code = 0x03 (read holding register)
// - address to read = word 33
// - data words to read = 6
// - token to match the response with the request.
//
  err = MB.addRequest(Token++, 247, READ_INPUT_REGISTER  , 4912,1);
            
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// The output on the Serial Monitor will be (depending on your Modbus the data will be different):
//      __ OK __
//      Response: serverID=1, FC=3, Token=00000457, length=15:
//      01 03 0C 60 61 62 63 64 65 66 67 68 69 6A 6B

}




void loop() {
}
```
the response is the following
![Capture](https://user-images.githubusercontent.com/24602056/170588933-510f3562-3345-47ac-a379-c1e093076957.PNG)

**Comment by Miq1** [comment, marked as answer]
Hi! I took the liberty to put your sketch in code tags to make it better readable.
At a first glance the code looks okay, so we may have to look elsewhere. Is the register you are attempting to read in fact an input and no hold register? It totally depends on the device how register types and addresses are interpreted. This of course should be mentioned in the data sheet somewhere. Second, it is quite common for devices to give register addresses counted from 1 on, while requiring requests being made with address-1, starting at 0.
So the first things I would try are READ_HOLD_REGISTER as the function code and in a second experiment READ_INPUT_REGISTER and address 4911.

**Reply by jagonzalezga** [reply]
i got it, so the sentence get(index, word) only can works with the define READ_INPUT_REGISTER , and te respnse is following
![image](https://user-images.githubusercontent.com/24602056/170728355-8b4fa436-226e-45a9-a5d1-37198765327c.png)
```
#include <Arduino.h>

// Include the header for the ModbusClient RTU style
#include "ModbusClientRTU.h"
#define rtsPin  22
// Create a ModbusRTU client instance
// In my case the RS485 module had auto halfduplex, so no second parameter with the DE/RE pin is required!
//ModbusClientRTU MB(Serial2);
ModbusClientRTU MB(Serial2, rtsPin);  // use rtsPin to toggle DE/RE in half-duplex

// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the message data and length of it, 
// plus a user-supplied token to identify the causing request
void handleData(ModbusMessage response, uint32_t token) 
{ uint32_t words;
  
  


  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d,Index=%u,words=%u \n", response.getServerID(), response.getFunctionCode(), token, response.size(),response.get(4,words),words);
  for (auto& byte : response) 
  {
    Serial.printf("%02X ", byte);
  }
  
  Serial.println("");
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", (int)me, (const char *)me);
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Set up Serial2 connected to Modbus RTU
// (Fill in your data here!)
  Serial2.begin(115200, SERIAL_8E1, GPIO_NUM_17, GPIO_NUM_18);

// Set up ModbusRTU client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task
  MB.begin();




// If something is missing or wrong with the call parameters, we will immediately get an error code 
// and the request will not be issued
  uint32_t Token = 1111;


  Error err;


// Create request for
// (Fill in your data here!)
// - server ID = 1  //masico 247
// - function code = 0x03 (read holding register)
// - address to read = word 33
// - data words to read = 6
// - token to match the response with the request.
//
  err = MB.addRequest(Token++, 247, READ_INPUT_REGISTER  , 4900,15);
            
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// The output on the Serial Monitor will be (depending on your Modbus the data will be different):
//      __ OK __
//      Response: serverID=1, FC=3, Token=00000457, length=15:
//      01 03 0C 60 61 62 63 64 65 66 67 68 69 6A 6B

}

```


void loop() {
}`

**Reply by jagonzalezga** [reply]
Thank you all for helping me, the mistake was that I did not understand the library well, but with your help I was able to achieve the objective.
so here is the answer to the problem

```cpp
Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d,Index=%u,words=%u \n", response.getServerID(), response.getFunctionCode(), token, response.size(),response.get(2,addr),addr);
  for (auto& byte : response) 
  {
    Serial.printf("%02X ", byte);
  }
  
  Serial.println("");
}
```
In these sentences I get the values ​​of register 4900 and I only want register 10, which returns the slave address value.

then 
`err = MB.addRequest(Token++, 247, READ_INPUT_REGISTER  , 4900,10);`

in this statement I make the request to read the register 4900 of the slave with address 247 (hex = 0xF7), and I want you to return only the first 10 values, then the last value corresponds to address 4910 which results in 0xF7 which is a 247 decimal.

![image](https://user-images.githubusercontent.com/24602056/170734011-ede5c5f1-c0ae-4165-afa1-42213a3fe78f.png)

In addition, they are two-byte responses, in total there are 20 bytes eliminating the first three

**Reply by Miq1** [reply]
Uhm, I do not quite understand what you are trying to explain, but good if that is working for you.
If you are reading 10 **registers** starting at 4900, you of course will get 20 bytes, because a Modbus register always is 16 bits wide, hence 2 bytes. To only read the tenth register you can request a single register from address 4909.
The ``get()`` function is totally agnostic to whatever the message may consist of - it will jump to the given ``index`` byte offset from the start and pick as much bytes as are needed for the variable you gave in the call.
I would suggest to have some more looks at the documentation and example sketches.

**Comment by bertmelis** [comment]
You also have to connect DE. You can tie the DE and RE pins together.

**Reply by Miq1** [reply]
The response was well-formed despite the DE pin not connected, it just gave back a value of zero.

**Reply by bertmelis** [reply]
👍  I see.

**Reply by jagonzalezga** [reply]
Thanks for answering, the cable is connected, only the cable is white and cannot be seen in the photo

**Reply by Miq1** [reply]
Now, that you wrote it... 😁 
Any progress? The device _is_ responding as we saw, so something in the logics must be wrong.

**Reply by jagonzalezga** [reply]
the device did work by changing these two lines in the code

Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d,Index=%u,words=%u \n", response.getServerID(), response.getFunctionCode(), token, response.size(),response.get(2,words),words);

err = MB.addRequest(Token++, 247, READ_INPUT_REGISTER  , 4900,15);
